src/shared/switch-root.c

```diff
@@ -149,9 +147,44 @@ int switch_root(const char *new_root,
                 if (r > 0) /* If it is already mounted, then do nothing */
                         continue;
 
-                r = mount_nofollow_verbose(LOG_ERR, transfer->path, chased, NULL, mount_flags, NULL);
-                if (r < 0)
-                        return r;
+                if (FLAGS_SET(flags, SWITCH_ROOT_RECURSIVE_RUN)) {
+                        /* On recursive runs, first create a bind mount with MOVE_MOUNT_BENEATH, and then MS_MOVE
+                         * the original mount point to the new root. This ensures that the mount point in the final
+                         * destination is the original mount, but also ensures that the mount is available in the old
+                         * root during the switch root. The latter is true because when we MS_MOVE, the bind mount is revelead. */
+                        _cleanup_close_ int mnt_fd = -EBADF, target_fd = -EBADF;
+
+                        target_fd = open(transfer->path, O_PATH|O_NOFOLLOW|O_CLOEXEC);
+                        if (target_fd < 0)
+                                return log_error_errno(errno, "Failed to open '%s': %m", transfer->path);
+
+                        mnt_fd = open_tree(-EBADF, transfer->path, OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC);
+                        if (mnt_fd < 0)
+                                return log_error_errno(errno, "Failed to open_tree '%s': %m", transfer->path);
+
+                        r = RET_NERRNO(mount_setattr(mnt_fd, "", AT_EMPTY_PATH,
+                                                     &(struct mount_attr) { .propagation = 0 },
+                                                     MOUNT_ATTR_SIZE_VER0));
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to set mount attrs: %m");
```

Comment:
sometimes you use RET_NERRNO() and sometimes you don't here. Both is fine, but it's very confusing.